src/proxy.rs

```diff
@@ -502,11 +527,129 @@ fn handle_server_event(
             );
 
             // TODO(#145): Fix unwrap
-            let handle = client_to_proxy.idle_reject(status).unwrap();
+            let handle = client_to_proxy.state.idle_reject(status).unwrap();
             trace!(role = "p2c", ?handle, "idle_reject");
         }
         client::Event::IdleDoneSent { handle } => {
             trace!(role = "p2s", ?handle, "--->");
         }
     }
 }
+
+struct ImapState<S> {
+    input_role: &'static str,
+    input_direction: &'static str,
+    input_color: Option<Color>,
+    input_fragmentizer: Fragmentizer,
+    output_role: &'static str,
+    output_direction: &'static str,
+    output_color: Option<Color>,
+    output_fragmentizer: Fragmentizer,
+    state: S,
+}
+
+impl ImapState<Client> {
+    pub fn client(state: Client) -> Self {
+        Self {
+            input_role: "s2p",
+            input_direction: "<--|",
+            input_color: Some(Color::Blue),
+            input_fragmentizer: Fragmentizer::without_max_message_size(),
+            output_role: "p2s",
+            output_direction: "--->",
+            output_color: None,
+            output_fragmentizer: Fragmentizer::without_max_message_size(),
+            state,
+        }
+    }
+}
+
+impl ImapState<Server> {
+    pub fn server(state: Server) -> Self {
+        Self {
+            input_role: "c2p",
+            input_direction: "|-->",
+            input_color: Some(Color::Red),
+            input_fragmentizer: Fragmentizer::without_max_message_size(),
+            output_role: "p2c",
+            output_direction: "<---",
+            output_color: None,
+            output_fragmentizer: Fragmentizer::without_max_message_size(),
+            state,
+        }
+    }
+}
+
+impl<S: imap_next::State> imap_next::State for ImapState<S> {
+    type Event = S::Event;
+    type Error = S::Error;
+
+    fn enqueue_input(&mut self, bytes: &[u8]) {
+        self.input_fragmentizer.enqueue_bytes(bytes);
+        while let Some(fragment_info) = self.input_fragmentizer.progress() {
+            handle_fragment_event(
+                self.input_role,
+                self.input_direction,
+                self.input_color,
+                &self.input_fragmentizer,
+                fragment_info,
+            );
+        }
```

Comment:
The idea here is to follow along the bytes with a separate `Fragmentizer`, yes? This makes me a bit unconfortable... Not because of performance, but more about the possibility that things -- due to some reason -- might get out-of-sync. Then, we would have confusing logs and not really a way to detect it.

Reply:
I agree with you. But! It's a really simple solution. And that's a strong argument, too.

Do you have any alternative ideas?

The only alternative I can think of is moving the log statements into `imap-next`. But there the log statements would a be a little bit more complicated. And it would be ugly that `imap-proxy` relies on log statements inside of `imap-next`. It would be more complicated to implement and ensure the same format and color. But maybe that's okay? Because fragment logging is an advanced feature for expert users?

For me it's the other way around. Fragment logging is more useful than the other stuff logged by the proxy.

Another argument: Normal users of `imap-next` might also be interested in fragment logging. If someone reports a bug, we could ask this person to give us the fragment logs.

What do you think?

Reply:
Agree to agree :-)

Before your implementation, the [`io/read/raw`](https://github.com/duesee/imap-next/blob/29072279a68c299d679c5712ec03c69fceff801f/src/stream.rs#L210) messages were most useful for me.

(I used `RUST_LOG=proxy=trace` and not `RUST_LOG=trace` and thought they disappeared in some refactor -- should adapt the README. These come from `imap-next` and don't have the same color and style.)

We could generally emit and catch all logs from `imap-{proxy,next,codec,types}` using a custom subscriber. This would solve the style issue. We could even wite a small ratatui and provide different views by multiplexing the logs into ratatui panes. I did something similar in our [`fake-mail-proxy`](https://github.com/Email-Analysis-Toolkit/fake-mail-server/blob/main/src/log.rs#L46). (The code is aweful. I didn't understand `tracing` at that time and just hacked around to make it work quickly.)

Thinking about it... I think it's worth to keep the fragment logging (as it is). It is something I want to use already.

By the way: The raw messages give you an intuition about the implementation of the client/server. Some email clients somehow flush the TCP buffers before and after "\r\n" which results in traces like ...

```
io/read/raw "A LOGIN user pass"
io/read/raw "\r\n"
```

I found this kinda interesting, too.

Let's try to merge this :-)

---

Let me dream ...

```rust
for event in client.progress() {
    match event {
        Event::Io(...) => { /* bytes received */ },
        Event::Tag(tag) => { /* tag found */ },
        Event::Line(line) => { /* line found */ },
        Event::Interrupt(question) => { /* e.g. need to accept literal */ },
        Event::Literal(literal) => { /* literal found */ },
        Event::Message(message) => { /* complete message found */ },
    }
}
```

Reply:
We want to keep it for now. In the long run we want to move as much logging to `imap-next` as possible and refactor `imap-proxy` to a log collecting TUI.